### PR TITLE
prov/efa: Enable SM2 provider as peer provider

### PR DESCRIPTION
**Pull request by a-szegel**
Implement FI_EFA_USE_SM2 environment variable to switch between SHM and SM2.

This cannot merge until https://github.com/ofiwg/libfabric/pull/8808 b/c SM2 in main doesn't implement peer provider.

@sunkuamzn is on vacation, so this updates his changes here: https://github.com/ofiwg/libfabric/pull/8791

### CODE REVIEW

File: prov/efa/src/efa_av.c [hunk 1]
```diff
@@ -327,12 +327,12 @@ int efa_conn_rdm_init(struct efa_av *av, struct efa_conn *conn)
 		 * The shm provider supports FI_AV_USER_ID flag. This flag
 		 * associates a user-assigned identifier with each av entry that is
 		 * returned with any completion entry in place of the AV's address.
-		 * In the fi_av_insert_call below, the &peer->shm_fiaddr is both an input
+		 * In the fi_av_insert call below, the &peer->shm_fiaddr is both an input
```
Comment (shijin-aws) [inline prov/efa/src/efa_av.c:330]: what changed in this line?

File: prov/efa/src/efa_domain.c [hunk 2]
```diff
@@ -133,7 +133,11 @@ static int efa_domain_init_rdm(struct efa_domain *efa_domain, struct fi_info *in
 	efa_shm_info_create(info, &efa_domain->shm_info);
 
 	if (efa_domain->shm_info) {
-		assert(!strcmp(efa_domain->shm_info->fabric_attr->name, "shm"));
+		if (rxr_env.use_sm2) {
```
Comment (shijin-aws) [inline prov/efa/src/efa_domain.c:136]: can we use a char * shm_provider_name to refer to either shm or sm2 and use a single assert here.

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c [hunk 3]
```diff
@@ -361,7 +361,12 @@ int efa_rdm_ep_open(struct fid_domain *domain, struct fi_info *info,
 		if (ret)
 			goto err_destroy_base_ep;
 
-		assert(!strcmp(efa_domain->shm_info->fabric_attr->name, "shm"));
+		if (rxr_env.use_sm2) {
+			assert(!strcmp(efa_domain->shm_info->fabric_attr->name, "sm2"));
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_ep_fiops.c:365]: similar comment as above.

File: prov/efa/src/efa_av.c (hunk 1 above)
Comment (a-szegel) [reply prov/efa/src/efa_av.c:330]: underscore: `fi_av_insert_call` to `fi_av_insert call`

... Sai had it captured in his PR.

**Review by shijin-aws — APPROVED** [review]

**Comment by a-szegel** [comment]
@wzamazon

File: prov/efa/src/efa_shm.c [hunk 4]
```diff
@@ -98,12 +98,26 @@ void efa_shm_info_create(const struct fi_info *app_info, struct fi_info **shm_in
...
+	} else {
+		shm_provider = "shm";
+	}
+
 	shm_hints = fi_allocinfo();
-	shm_hints->caps = FI_MSG | FI_TAGGED | FI_RECV | FI_SEND | FI_READ
-			   | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE
-			   | FI_MULTI_RECV | FI_RMA | FI_SOURCE;
-	shm_hints->domain_attr->av_type = FI_AV_TABLE;
+	shm_hints->caps = app_info->caps;
```
Comment (wzamazon) [inline prov/efa/src/efa_shm.c:109]: I think we need to remove `FI_REMOTE_COMM`

File: prov/efa/src/rdm/rxr_env.c [hunk 5]
```diff
@@ -228,6 +230,8 @@ void rxr_env_define()
 			"Enables fork support and disables internal usage of huge pages. Has no effect on kernels which set copy-on-fork for registered pages, generally 5.13 and later. (Default: false)");
 	fi_param_define(&efa_prov, "runt_size", FI_PARAM_INT,
 			"The maximum number of bytes that will be eagerly sent by inflight messages uses runting read message protocol (Default 307200).");
+	fi_param_define(&efa_prov, "use_sm2", FI_PARAM_BOOL,
+			"Use the experimental shared memory provider SM2.");
```
Comment (wzamazon) [inline prov/efa/src/rdm/rxr_env.c:234]: For intra node communication

**Review by wzamazon — APPROVED** [review]

File: prov/efa/src/efa_shm.c [hunk 6]
```diff
@@ -98,12 +98,27 @@ void efa_shm_info_create(const struct fi_info *app_info, struct fi_info **shm_in
...
+	shm_hints->caps &= ~FI_REMOTE_COMM;
```
Comment (shijin-aws) [inline prov/efa/src/efa_shm.c:110]: wait, is this cap is hints->caps or hints->domain_attr->caps?

Comment (shijin-aws) [reply prov/efa/src/efa_shm.c:110]: See L122

Comment (shijin-aws) [reply prov/efa/src/efa_shm.c:110]: Looks like putting in hints->caps is fine. This is how ompi mtl ofi does
```
hints->caps               = FI_TAGGED | FI_LOCAL_COMM | FI_REMOTE_COMM | FI_DIRECTED_RECV;
```

Comment (shijin-aws) [reply prov/efa/src/efa_shm.c:110]: I am ok as long as you have tested that shm can be picked up.

Comment (a-szegel) [reply prov/efa/src/efa_shm.c:110]: Yes I did test this and verify SHM can still be picked up (with fabtests + osu benchmarks on OMPI 5).

**Pull request merged**